### DISCUSSION TITLE
Order Management: How to auto close position on TP or SL hit?

### DISCUSSION THREAD

**Discussion by texasraj** [kernc/backtesting.py#824] [Q&A]
Over the past week, I have been scratching about the order management here.. I see a big discrepancy in the orders.. especially, closing the positions.. 

Basically, the strategy is not closing the positions though SL or TP hit.. ?

**Comment by kernc** [comment]
> does it close the position automatically when it hits SL or TP?

Yes, the both orders (SL and TP, as well as the trade they counter) are closed automatically when hit.

> Second, how to cancel open orders waiting to be executed based on the limit price? How are the limit orders closed here?

[`Order.cancel()`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Order.cancel) is the way to go. However, when you're iterating over orders, be mindful of [`Order.is_contingent`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Order.is_contingent) flag:
```py
for order in self.orders:
    if not order.is_contingent:  # Skip SL/TP orders
        order.cancel()
```

**Reply by makovez** [reply]
How can I close an order that has sl and tp if it didn't reach the tp in a certain amount of candles?

**Reply by kernc** [reply]
@makovez See e.g. https://github.com/kernc/backtesting.py/issues/502#issuecomment-1374887567.

**Reply by makovez** [reply]
Thanks. But does .cancel() calculate the pl? I need to close the trade, not to cancel it.

**Reply by kernc** [reply]
Have you seen https://github.com/kernc/backtesting.py/discussions/265#discussioncomment-399324 or https://github.com/kernc/backtesting.py/discussions/505#discussioncomment-1468573? Closing a trade will (should) cancel its contingent orders.

**Comment by makovez** [comment]
Yes thank you. It works.

Il gio 19 gen 2023, 15:39 kernc ***@***.***> ha scritto:
